formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt

```diff
@@ -97,7 +127,18 @@ internal open class ProtobufDecoder(
 
     override fun beginStructure(descriptor: SerialDescriptor): CompositeDecoder {
         return when (descriptor.kind) {
-            StructureKind.LIST -> RepeatedDecoder(proto, reader, currentTagOrDefault, descriptor)
+            StructureKind.LIST -> {
+                val tag = currentTagOrDefault
+                return if (tag != MISSING_TAG && this.descriptor != descriptor && this.descriptor.kind == StructureKind.LIST) {
```

Comment:
I think `this.descriptor.kind == StructureKind.LIST` the most important and rare clause so it should come first. Also, a comment e.g. '// decoding nested list' would explain things better, because it's hard to understand why this `if` needed at all